### ISSUE TITLE
Specification evaluator improvements

### ISSUE THREAD

**Issue by fiseni** [ardalis/Specification#28]
Since this grew to be a separate construct, then let just do it properly. Lets make the methods non-static, and define an interface. If you decide to go with infrastructure as described in issue #10 and include the repository implementation in the package, then the interface would be injected in the constructor.
The included repository actually would be abstract, and the users will have to inherit it, so they can use their own DbContext. It might be something as below

```
    public abstract class Repository<T> : IRepository<T> where T : BaseEntity, IAggregateRoot
    {
        private readonly DbContext dbContext;
        protected readonly ISpecificationEvaluator<T> specificationEvaluator;

        public Repository(DbContext dbContext)
        {
            this.dbContext = dbContext;
            this.specificationEvaluator = new SpecificationEvaluator<T>();
        }

        public Repository(DbContext dbContext, ISpecificationEvaluator<T> specificationEvaluator)
        {
            this.dbContext = dbContext;
            this.specificationEvaluator = specificationEvaluator;
        }

        // Interface implementation
        // We should define all the methods as virtual, or most of them.

    }

    public class MyRepository<T> where T : BaseEntity, IAggregateRoot
    {
        private readonly MyDbContext myDbContext;

        public MyRepository(MyDbContext myDbContext)
            : base(myDbContext)
        {
            this.myDbContext = myDbContext;
        }

        public MyRepository(MyDbContext myDbContext, ISpecificationEvaluator<T> specificationEvaluator)
            : base(myDbContext, specificationEvaluator)
        {
            this.myDbContext = myDbContext;
        }

        // Not required to implement anything. Override or add additional functionalities.
    }
```

If the user has not defined their own implementation for the SpecificationEvaluator, then the default one will be instantiated. And yet again, they can implement their own. This structure offers much more flexibility.

**Comment by ardalis** [comment]
I'm not 100% clear on this - are you saying we would ship an abstract base Repository type that all users would have to inherit in order to make use of Specification? Seems like a big request compared to only being a general purpose specification library that can be used with any Repository that uses a certain convention.

**Comment by fiseni** [comment]
Oh no, I didn't mean to enforce it. The idea is the following:
- Users should be able to create specifications utilizing the BaseSpecification
- Users can create their own repositories and use instance of SpecificationEvaluator to evaluate the specs into IQueryable
- Users can leverage from the provided abstract repository (if they want), which contains set of standard generic methods.
- If users want to make use of the provided repository but want to utilize their own evaluator, they can inject instance of ISpecificationEvaluator.

I tried to describe everything and give few samples in the Readme of the repo.

**Comment by ardalis** [comment]
Ok that sounds perfect, thanks.

**Comment by fiseni** [comment]
This is addressed with PR #30 and #31
